text/0000-hidden-impls.md

```diff
@@ -0,0 +1,878 @@
+# Draft RFC: Hidden implementations
+
+- Feature Name: `hidden_impl`
+- Start Date: 2018-06-24
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow a visibility modifier on a `trait` implementation such that:
+
++ it is public wrt. coherence and overlap
++ it is not usable in contexts where it is *hidden* (not visible)
+
+An example:
+
+```rust
+#[derive(Clone)]
+struct Foo(..);
+
+crate impl Copy for Foo {} // Can only be used in the current crate!
+```
+
+# Motivation
+[motivation]: #motivation
+
+First, let's assume that we have a trait `Property` defined somewhere.
+This trait, which doesn't have to be inside your crate, is also visible
+outside of it. It also happens that your crate has defined and exported
+a type called `Thing`. The specifics of `Property` and `Thing` are not
+interesting for our purposes, but they exist.
+
+You now want to, define an implementation of `Property` for `Thing`.
+But for some reason, you are not willing, or able, to expose this implementation
+outside your crate and make it part of your semantic versioning guarantees.
+
+An example of such a scenario is defining a new-type wrapper `TestRng` around
+`XorShiftRng` of the `rand` crate. We want to define an implementation of
+`SeedableRng` for `TestRng` to take advantage of the logic of `FromEntropy`
+which is automatically provided for us when `Rng: SeedableRng`.
+But we also want to avoid exposing `SeedableRng` as a guarantee to the
+users of our library (`proptest`) right now. The allure of getting the logic
+provided when defining `SeedableRng` could also have caused us to prematurely
+and mistakenly provide a guarantee we were not ready to provide.
+
+Another example use-case which [@nikomatsakis](https://github.com/nikomatsakis)
+has provided is wanting to have the convenience of `#[derive(Default)]`,
+but not wanting to commit to there being a default value.
+
+However, we have no way eat our cake and keep it. We can't segregate who
+has access to what implementations by visibility.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The basic idea
+
+The problem set out by the [motivation] is solved by allowing you to
+hide away implementations which you, for whatever reason, are not willing or
+able to expose to users of your library. This RFC solves that with a simple
+solution: *visibility modifiers on implementations.*
+
+Instead of writing:
+
+```rust
+// crate A:
+
+impl Property for Thing { .. }
+```
+
+we can now write:
+
+```rust
+// crate A:
+
+crate impl Property for Thing { .. }
+```
+
+where `crate`, or `pub(crate)`, is the familiar visibility
+modifier for the current crate.
+
+## The effect of hiding
+
+### Error E0277
```

Comment:
Note that there's a non-trivial interaction with *macros* here; currently, "privacy hygiene" is not a thing.

Reply:
Could you elaborate a bit here? The way I'm thinking is that if we have a macro with say:

```rust
macro_rules! mac {
    () => {
        fn return_property() -> impl Property {
            make_a_thing() // ← error[E0277]
        }
    }
}
```

which is then called at some site, then the visibility context used is that of the call site of `mac!()`.
The same should be the case with a macro defining an implementation; it too is visible according to the call-site and not the definition site of the macro.